picard/util/imagelist.py

```diff
@@ -112,3 +113,7 @@ def get_types_dict(self):
                     continue
             types_dict[image_types] = image
         return types_dict
+
+    def wait_for_image_processing(self):
```

Comment:
Ideally this should be callback based and not wait with polling.

But currently the threading tools provided by Picard do not make that easy. The most common case in Picard is either running a single thread then callback or in some cases a sequential queue of things to run in a thread one after another.

I think we need some general way to run code after a bunch of threads has finished. @zas, what do you think?

In any way we could do it with the wait and status polling as part of this PR and refactor to a different mechanism later.

Reply:
IMO there are a bunch of things that should be provided as general methods, and an improved threading approach is one of them, generalised caching being another.

Reply:
> Unless it is impossible to achieve, polling should be avoided at all costs IMO.

Can you elaborate on this?
What alternatives do you suggest?

Reply:
Sorry - bit it is so long since I looked at the multi-threading code that I am unable to make any suggestions.

From memory of c. a decade and a half ago, I felt that the way that stuff was passed to and from a thread was wrong and that queues (and possibly signals) would be a better architecture than the current one (or rather the one at that time).

Plus queues gives a common approach for both multithreading **and** multi-processing.

As for generic methods, examining the code of existing plugins for stuff that is plumbing (that is making an individual plugin much more complicated than it really should need to be and / or which is duplicated across >= 2 plugins) rather than direct functionality might well be a decent way to establish what generic code is needed.